File: onboard/defnodes/DriveInterface/external/DEFCOM/src/ChannelMulticast.cpp
#include "comms/UnixWrapper.hpp"
#include "comms/GenericNetWrapper.hpp"
#include "comms/TCPWrapper.hpp"
#include "ChannelMulticast.hpp"
#include <filesystem>
#include <thread>
#include <chrono>

//Server Stuff
MulticastPublisher::MulticastPublisher(std::string defComFile){
    //Load the comms definitions
    this->definition = loadConfFile(defComFile);

    //Open the TCP server
    this->tcpServer = newTCPServer(this->definition.ResolvedIP, this->definition.NumericPort);

    //Open the Unix server
    this->unixServer = newUnixServer("/tmp/" + this->definition.Name);

    //Start the threads
    acceptorThreadTCP = std::thread(&MulticastPublisher::acceptorTCP, this);
    acceptorThreadUnix = std::thread(&MulticastPublisher::acceptorUnix, this);

}

MulticastPublisher::~MulticastPublisher(){
    //Join all the threads
    //acceptorThreadTCP.join();
    //acceptorThreadUnix.join();

    //Clean up dangling connections
    std::cout << "Shutting Down" << std::endl;
    for (size_t i = 0; i < clientConnections.size(); ) {
        delete clientConnections[i];
    }
    std::cout << "Destroy Clients" << std::endl;

}

void MulticastPublisher::acceptorTCP(){
    while(true){
        //Accept the connection
        TCPServerCon* client = new TCPServerCon(this->tcpServer);

        //Log the connection
        std::cout << "Accepted TCP connection" << std::endl;

        this->spawnClient(client);
    }
}

void MulticastPublisher::acceptorUnix(){
    while(true){
        //Accept the connection
        UnixServerCon* client = new UnixServerCon(this->unixServer);

        //Log the connection
        std::cout << "Accepted Unix connection" << std::endl;

        this->spawnClient(client);
    }
}

void MulticastPublisher::spawnClient(GenericNetWrapper* client){
    //Lock the acceptor mutex
    this->acceptMutex.lock();

    //Spawn the client
    clientConnections.push_back(client);
    clientThreads.emplace_back(&MulticastPublisher::clientProcess, this, client, idCounter);

    //Increment the client ID counter
    idCounter++;

    //Clean up dead threads
    for (size_t i = 0; i < clientConnections.size(); ) {
        if (!clientConnections[i]->report().Alive) {
            //Free the client
            delete clientConnections[i];

            //Join the thread
            clientThreads[i].join();

            //erase both at the same index
            clientConnections.erase(clientConnections.begin() + i);
            clientThreads.erase(clientThreads.begin() + i);

        } else {
            ++i;
        }
    }

    //Release the acceptor mutex
    this->acceptMutex.unlock();    
}

void MulticastPublisher::clientProcess(GenericNetWrapper* client, int clientID){
    //Create my entry in the send buffers
    this->sendMutex.lock();
    sendBuffers.emplace(clientID, std::vector<MessageStructure>{});
    this->sendMutex.unlock();

    while (client->report().Alive){
        while (sendBuffers[clientID].size() == 0){
            //Wait 0.1s
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        if (sendBuffers[clientID].size() > 0){
            this->sendMutex.lock();
            bool result = client->senddat(sendBuffers[clientID].back().getDecodeBuffer(), sendBuffers[clientID].back().totalSize);
            sendBuffers[clientID].pop_back();
            this->sendMutex.unlock();

            if (!result){
                break;
            }

        }

    }
    client->closeCon();
    std::cout << "Client closed" << std::endl;

    //Destroy my entry in the send buffers
    this->sendMutex.lock();
    this->sendBuffers.erase(clientID);
    this->sendMutex.unlock();

}


MessageStructure MulticastPublisher::getNewMessageObject(){
    //Create a new request object
    MessageStructure request(this->definition.RequestMessageFormat);

    return request;
}


void MulticastPublisher::publish(MessageStructure requestData){
    //Lock the mutex
    this->sendMutex.lock();

    //Add the message to each of the send buffers
    for (auto& clientID : sendBuffers){
        clientID.second.emplace_back(requestData);
    }
    
    //Unlock the mutex
    this->sendMutex.unlock();

}


// Client Stuff
MulticastSubscriber::MulticastSubscriber(std::string defComFile){
    //Load the comms definitions
    this->definition = loadConfFile(defComFile);

    //Check if the unix file exists first
    if (std::filesystem::exists("/tmp/" + this->definition.Name)) {
        this->con = new UnixClientCon("/tmp/" + this->definition.Name);
        std::cout << "Connected Unix to " << this->definition.Name << std::endl;
    }
    else {
        this->con = new TCPClientCon(this->definition.ResolvedIP, this->definition.NumericPort);
        std::cout << "Connected TCP to " << this->definition.ResolvedIP << ":" << this->definition.NumericPort << std::endl;
    }

}


MulticastSubscriber::~MulticastSubscriber(){
    //Close the connection
    this->con->closeCon();
    delete this->con;
}



MessageStructure MulticastSubscriber::subscribe(){

    // Get the response
    unsigned char* message = this->con->getdat(this->definition.RequestMessageFormat.totalSize);

    //Decode the response
    MessageStructure response(this->definition.RequestMessageFormat);
    if (message != nullptr){
        response.setDecodeBuffer(message);
    }

    return response;
}

File: basestation/Backend/DEFCOM/ChannelMulticast.py
import threading
from ._FlexibleMessageStructure import MessageStructure
from ._DefComParser import ConnectionSpecification as _ConnectionSpecification, loadConfFile as _loadConfFile
from ._TCPWrapper import clientTCPCon as _clientTCPCon, serverTCPCon as _serverTCPCon, newTCPServer as _newTCPServer
from ._UNIXWrapper import clientUnixCon as _clientUnixCon, serverUnixCon as _serverUnixCon, newUnixServer as _newUnixServer
import time
import os

class MulticastPublisher:
    #Initialise with the message structure definition (defcom file) and a function pointer
    #The function must be of the type:  void function(MessageStructure request, MessageStructure response)
    def __init__(self, defComFile: str):
        self._definition: _ConnectionSpecification = _loadConfFile(defComFile)

        #A mutex for the sending thread
        self._sendMutex = threading.Lock()

        #Open the tcp server on the decoded port and address
        self._tcpServer = _newTCPServer(self._definition.ResolvedIP, self._definition.NumericPort)

        #Open a named socket as well
        self._unixServer = _newUnixServer('/tmp/' + self._definition.Name)

        #The acceptor thread for TCP
        self._acceptorThreadTCP = threading.Thread(target=self._acceptorTCP)

        #The acceptor thread for Unix
        self._acceptorThreadUnix = threading.Thread(target=self._acceptorUnix)

        #The client thread array
        self._clientThreads = []
        
        #Outgoing buffer
        self._outgoingQueues = {}
        self.idCounter = 0

        self._acceptorThreadTCP.start()
        self._acceptorThreadUnix.start()


    #Get a message to be filled in
    def getNewMessageObject(self) -> MessageStructure:
        return self._definition.RequestMessageFormat.clone()

    def publish(self,requestData: MessageStructure):
        #Add the message to the outgoing queue
        with self._sendMutex:
            for key in self._outgoingQueues:
                self._outgoingQueues[key].append(requestData)

    # Internal Client creator
    def _spawnClient(self, client):
        #Start a new thread to handle it
        with self._sendMutex:
            clientNewThread = threading.Thread(target=self._clientProcess, args=(client,self.idCounter))
            clientNewThread.start()
            self._clientThreads.append(clientNewThread)
            self.idCounter += 1

            #Clean up dead clients
            ToCull = []
            for i in self._clientThreads:
                if not i.is_alive():
                    ToCull.append(i)
            for i in ToCull:
                self._clientThreads.remove(i)

    # Internal thread that accepts client connections and spins off new threads for them
    def _acceptorTCP(self):
        while True:
            #Accept a client
            client = _serverTCPCon(self._tcpServer)
            print("Accepted TCP Client {}".format(client.info["Address"]))

            self._spawnClient(client)

    def _acceptorUnix(self):
        while True:
            #Accept a client
            client = _serverUnixCon(self._unixServer)
            print("Accepted UNIX Client {}".format(client.info["Address"]))

            self._spawnClient(client)

            

            
            
    #The client handler thread
    def _clientProcess(self, con, myID):
        # Basically just wait for requests and provide responses
        Running = True
        with self._sendMutex:
            self._outgoingQueues[myID] = []

        while Running:
            if con.info["Alive"]:
                if len(self._outgoingQueues[myID]) > 0:
                    with self._sendMutex:
                        while len(self._outgoingQueues[myID]) > 0:
                            #Send the data
                            if not con.senddat(self._outgoingQueues[myID][0].getDecodeBuffer()):
                                Running = False

                            #Next message
                            self._outgoingQueues[myID] = self._outgoingQueues[myID][1:]
                else:
                    time.sleep(0.1)

        con.close()
        print("Client Disconnected {} {}".format(con.info["Address"]))

        #Delete the queue
        with self._sendMutex:
            del self._outgoingQueues[myID]




class MulticastSubscriber:
    def __init__(self, defComFile: str):
        self._definition: _ConnectionSpecification = _loadConfFile(defComFile)

        #Connect to the server
        #If we are running on the same machine
        if self._definition.Name in os.listdir('/tmp'):
            self._con = _clientUnixCon('/tmp/' + self._definition.Name)
            print("Connected Unix to {}".format(self._definition.Name))
        else:
            self._con = _clientTCPCon(self._definition.ResolvedIP, self._definition.NumericPort)
            print("Connected TCP to {} port {}".format(self._definition.ResolvedIP, self._definition.NumericPort))
        

    def subscribe(self) -> MessageStructure:
        #Receive the response
        message = self._con.getdat(self._definition.RequestMessageFormat.totalSize)

        #Return the response
        response = self._definition.RequestMessageFormat.clone()
        
        if message:
            response.setDecodeBuffer(message)

        return response

    